domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+	}
+	machineUUID := result["machine_uuid"].(string)
+	return machineUUID, domain.Life(life), nil
+}
+
+// SetMachineBlockDevices sets the block devices visible on the machine.
+// Previously recorded block devices not in the list will be removed.
+func (st *State) SetMachineBlockDevices(ctx context.Context, machineId string, devices ...blockdevice.BlockDevice) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		machineUUID, life, err := getMachineInfo(ctx, tx, machineId)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if life == domain.Dead {
+			return errors.Errorf("cannot update block devices on dead machine %q", machineId)
+		}
+		existing, err := loadBlockDevices(ctx, tx, machineId)
+		if err != nil {
+			return errors.Annotatef(err, "loading block devices for machine %q", machineId)
+		}
+		if !blockDevicesChanged(existing, devices) {
+			return nil
+		}
+
+		if err := updateBlockDevices(ctx, tx, machineUUID, devices...); err != nil {
+			return errors.Annotatef(err, "updating block devices on machine %q (%s)", machineId, machineUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func updateBlockDevices(ctx context.Context, tx *sqlair.TX, machineUUID string, devices ...blockdevice.BlockDevice) error {
+	fsTypeQuery := `SELECT * AS &FilesystemType.* FROM filesystem_type`
+	fsTypeStmt, err := sqlair.Prepare(fsTypeQuery, FilesystemType{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var fsTypes []FilesystemType
+	if err := tx.Query(ctx, fsTypeStmt).GetAll(&fsTypes); err != nil {
+		return errors.Trace(err)
+	}
+	fsTypeByName := make(map[string]int)
+	for _, fsType := range fsTypes {
+		fsTypeByName[fsType.Name] = fsType.ID
+	}
+
+	if err := removeMachineBlockDevices(ctx, tx, machineUUID); err != nil {
+		return errors.Annotatef(err, "removing existing block devices for machine %q", machineUUID)
+	}
+
+	if len(devices) == 0 {
+		return nil
+	}
+
+	insertQuery := `
+INSERT INTO block_device (uuid, name, label, device_uuid, hardware_id, wwn, bus_address, serial_id, mount_point, size, filesystem_type_id, in_use)
+VALUES (
+    $BlockDevice.uuid,
+    $BlockDevice.name,
+    $BlockDevice.label,
+    $BlockDevice.device_uuid,
+    $BlockDevice.hardware_id,
+    $BlockDevice.wwn,
+    $BlockDevice.bus_address,
+    $BlockDevice.serial_id,
+    $BlockDevice.mount_point,
+    $BlockDevice.size,
+    $BlockDevice.filesystem_type_id,
+    $BlockDevice.in_use
+)
+`
+	insertStmt, err := sqlair.Prepare(insertQuery, BlockDevice{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	blockDevicesByUUID := make(map[utils.UUID]blockdevice.BlockDevice, len(devices))
+	for _, bd := range devices {
+		fsTypeID, ok := fsTypeByName[bd.FilesystemType]
+		if !ok {
+			return errors.NotValidf("filesystem type %q for block device %q", bd.FilesystemType, bd.DeviceName)
```

Comment:
Is this really not valid or is not found?

Reply:
Not Valid. The string fstype "blah" is not valid because it doesn't exist in the list of supported fs type names.